wgpu-hal/src/dynamic/instance.rs

```diff
@@ -0,0 +1,53 @@
+// Box casts are needed, alternative would be a temporaries which are more verbose and not more expressive.
+#![allow(trivial_casts)]
+
+use crate::{Capabilities, Instance, InstanceError};
+
+use super::{DynAdapter, DynResource, DynResourceExt as _, DynSurface};
+
+pub struct DynExposedAdapter {
+    pub adapter: Box<dyn DynAdapter>,
+    pub info: wgt::AdapterInfo,
+    pub features: wgt::Features,
+    pub capabilities: Capabilities,
+}
+
+pub trait DynInstance: DynResource {
+    unsafe fn create_surface(
```

Comment:
issue: We don't document a safety contract or adherence thereto for these methods. We should do this.